### ISSUE TITLE
Occasionally resources fail to load

### ISSUE THREAD

**Issue by kleinfreund** [disraptor/disraptor#5]
Requesting any Disraptor route multiple times in a Discourse production environment occasionally produces a 404 status code for an available resource.

Tested routes:

- https://disraptor.tira.io/tira9-client-web/static/css/bootstrap.min.css<br> → http://webis18.medien.uni-weimar.de:8084/tira9-client-web/static/css/bootstrap.min.css
- https://disraptor.tira.io/extern<br> → http://example.org/index.html

## Analysis

First of all, this only happens in the production environment. Requesting any document or resource over and over will sometimes lead to a 404 page while resolving fine otherwise.

There are two scenarios: Requesting a document and requesting a resource.

Disraptor distinquishes between the two. Disraptor documents (i.e. HTML resources) should be rendered inside Discourse. Therefore, they need to be requested via XmlHttpRequests from Discourse’s front end. All other resources (e.g. stylesheets, scripts, etc.) are handled by Discourse’s back end directly. Eventually, both types of resources are resolved by Disraptor’s proxy controller.

I tried reproducing this issue with both documents and resources and it turns out that both can result in this issue. The only difference is that for resources, the initial request fails (because it’s handled on the server side directly) while for documents, only the secondary request that is send from the Discourse client side fails.

This means that the Discourse proxy controller sometimes produces a 404 status code for the failing requests. However, the returned status code of the proxy is *always* exactly the status code that is returned by the proxy request.

In conclusion, somehow the proxy request produces a status code 404 for a resource that should always be available.

**Comment by kleinfreund** [comment]
I cannot reproduce this with my desktop computer (wired connection). Maybe the internet speed is a factor. At university, I was connected to eduroam which is not too stable.

I **can** reproduce this with my laptop (wireless connection) in the same network. The only difference is that my laptop is connected our home wi-fi and my desktop computer is connected to the router via wire.

**Comment by kleinfreund** [comment]
First of all, this only happens in the production environment.

- In the production environment, requesting http://disraptor.tira.io/tira9-client-web/static/css/bootstrap.min.css (or any other resource or document) over and over will sometimes lead to a 404 page while resolving fine otherwise.
- In the development environment, requesting http://localhost:3000/tira9-client-web/static/css/bootstrap.min.css always resolves fine.

There are two scenarios: Requesting a document and requesting a resource.

Disraptor distinquishes between the two. Disraptor documents (i.e. HTML resources) should be rendered inside Discourse. Therefore, they need to be requested via XmlHttpRequests from Discourse’s front end. All other resources (e.g. stylesheets, scripts, etc.) are handled by Discourse’s back end directly. Eventually, both types of resources are resolved by Disraptor’s proxy controller.

I tried reproducing this issue with both documents and resources and it turns out that both can result in this issue. The only difference is that for resources, the initial request fails (because it’s handled on the server side directly) while for documents, only the secondary request that is send from the Discourse client side fails.

This means that the Discourse proxy controller sometimes produces a 404 status code for the failing requests. However, the returned status code of the proxy is *always* exactly the status code that is returned by the proxy request.

In conclusion, somehow the proxy request produces a status code 404 for a resource that should always be available.

Next steps:

- Reproduce the issue with a resource that is not pointing to a target URL on Webis infrastructure.
- Try to verify these observations directly.

**Comment by kleinfreund** [comment]
After upgrading Discourse and Disraptor in the production environment, I can no longer reproduce this. I don’t trust this at all. :/

**Comment by potthast** [comment]
Indeed.

When looking closely, you can still see how the resources are loaded after the page is initially added to the Discourse context. When I hit F5 very often, I get 429 Too many requests, which is probably fine. But could this have been the case earlier, too?

A thought occurs: In the plugin, can we define a set of resources that are "immediately" added to all Disraptor pages, even before Ember asks the proxy for the actual page content, so that common resources like Bootstrap are available up front?
